sagemathinc/cocalc
# front-end tests for CoCalc in puppeteer

# Description

- This is proof-of-concept code. It works, but needs refactoring. See README.md for TODO list.
- This patch includes my last changes to the selenium suite, but I have moved all my front-end testing efforts to puppeteer. 
- All changes are under `src/test/puppeteer` and `src/test/selenium`, so there *should* be no impact to production code.
- Tests that work:
  - basic sign-in
  - open a .tex file and get word count
  - open a .ipynb and move an IntSlider() widget, verify change in readout
  - get api_key
- Platforms tested:
  - cocalc.com and test.cocalc.com
  - latest regular and no-agpl Docker images on UW server (removed after push to Docker Hub).
  - latest regular and no-agpl images from Docker Hub, on Pixelbook
- Tests run from a CoCalc project and from a git clone of cocalc sources on macOS.
  
# Testing Steps
1. See README.md. TL;DR:
    ```
    # check out the patch
    cd ~/cocalc/src/test/puppeteer
    npm ci # installs puppeteer, which will install local chromium binary about 100MB
    # create credentials.js file and setup test project per README.md
    # expect first one or two runs to fail if project has just been started
    node index.js -c path/to/credentials # omit ".js" suffix
    node widget.js -c path/to/credentials
    node api_key.js -c path/to/credentials
    ```
1.
1.

# Relevant Issues

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
